Sources/AsyncHTTPClient/AsyncAwait/HTTPClient+execute.swift

```diff
@@ -0,0 +1,173 @@
+                redirectState: redirectState,
+                request: preparedRequest,
+                response: response,
+                deadline: deadline,
+                logger: logger
+            )
+        }
+        return response
+    }
+
+    private func followRedirect(
+        redirectURL: URL,
+        redirectState: RedirectState,
+        request: HTTPClientRequest.Prepared,
+        response: HTTPClientResponse,
+        deadline: NIODeadline,
+        logger: Logger
+    ) async throws -> HTTPClientResponse {
+        var redirectState = redirectState
+        try redirectState.redirect(to: redirectURL.absoluteString)
+        let (method, headers, body) = transformRequestForRedirect(
+            from: request.url,
+            method: request.head.method,
+            headers: request.head.headers,
+            body: request.body,
+            to: redirectURL,
+            status: response.status
+        )
+        var newRequest = HTTPClientRequest(url: redirectURL.absoluteString)
+        newRequest.method = method
+        newRequest.headers = headers
+        newRequest.body = body
+        return try await self.executeAndFollowRedirectsIfNeeded(
+            newRequest,
+            deadline: deadline,
+            logger: logger,
+            redirectState: redirectState
+        )
+    }
+
+    private func executeWithoutFollowingRedirects(
+        _ request: HTTPClientRequest.Prepared,
+        deadline: NIODeadline,
+        logger: Logger
+    ) async throws -> HTTPClientResponse {
+        /// There is currently no good way to asynchronously cancel an object that is initiated inside the `body` closure of `with*Continuation`.
+        /// As a workaround we use `TransactionCancelHandler` which will take care of the race between instantiation of `Transaction`
+        /// in the `body` closure and cancelation from the `onCancel` closure  of `with*Continuation`.
+        actor TransactionCancelHandler {
+            private enum State {
+                case initialised
+                case register(Transaction)
+                case cancelled
+            }
+
+            private var state: State = .initialised
+
+            init() {}
+
+            func registerTransaction(_ transaction: Transaction) {
+                switch self.state {
+                case .initialised:
+                    self.state = .register(transaction)
+                case .cancelled:
+                    transaction.cancel()
+                case .register:
+                    preconditionFailure("transaction already set")
+                }
+            }
+
+            func cancel() {
+                switch self.state {
+                case .register(let bag):
+                    self.state = .cancelled
+                    bag.cancel()
+                case .cancelled:
+                    break
+                case .initialised:
+                    self.state = .cancelled
+                }
+            }
+        }
```

Comment:
I think we should probably pull this one out of this function... just put it on the file end with `private` access.